model-providers/watsonx/runtime/src/main/java/io/quarkiverse/langchain4j/watsonx/WatsonxStreamingChatModel.java

```diff
@@ -113,6 +98,111 @@ public void run() {
                         });
     }
 
+    @Override
+    public void generate(List<ChatMessage> messages, List<ToolSpecification> toolSpecifications,
+            StreamingResponseHandler<AiMessage> handler) {
+        TextGenerationRequest request = generateRequest(messages, toolSpecifications);
+        Context context = Context.of("response", new ArrayList<TextGenerationResponse>(), "toolExecution", false);
+
+        client.chatStreaming(request, version)
+                .subscribe()
```

Comment:
If the client calls back on an event loop, tool invocation will not be allowed to block.

Reply:
I'm sorry, I'm not following, can you explain?

Reply:
This callback will be invoked on an event loop, on thread you cannot block. It means that the tools you are going to invoke will be executed on that same thread, and thus won't be able to block. Typically, you will not be able to touch a database using a blocking API, or invoke another LLM using a  non reactive method.

Reply:
Thanks for the clarification. I now realise that the problem here is that all of the tool's functions have to be "asynchronous" because the thread can't be blocked.

Reply:
I think this is not enough, because the `core` module does not handle the Multi return type at the moment, or so it seems. Actually, there is still some work to be done here.

Reply:
I think there is a way forward if we can know whether the tools will be blocking or not.
1) If not blocking (returning an Uni), we may have to do some work in _core_.
2) If blocking, we can imagine switching to a worker thread before  the tools invocation and switching back after. 

2) can be done without 1) 

@geoand WDYT?

Reply:
Maybe the implementation is more complex than I thought. Looking at the code, `langchain4j` wants a `json` as the result of the tool call, so the `QuarkusToolExecutor` at line [66](https://github.com/quarkiverse/quarkus-langchain4j/blob/4eada1dcc8188ddfaa181468fc4fbf3e4621d842/core/runtime/src/main/java/io/quarkiverse/langchain4j/runtime/tool/QuarkusToolExecutor.java#L66) does the conversion of the result, but in this case we have a `Uni`, and of course that's something `jackson` doesn't like.

Reply:
If I change the `handleResult` in this way:

```java
private static String handleResult(ToolInvoker invokerInstance, Object invocationResult) {
        if (invokerInstance.methodMetadata().isReturnsVoid()) {
            return "Success";
        }

        if (invocationResult instanceof Uni uniInvocation) {
            return Json.toJson(uniInvocation.await().indefinitely());
        } else {
            return Json.toJson(invocationResult);
        }
    }
```

And I have a tool like that:

```java
@Tool("Execute the embedding of a string")
    public Uni<List<Float>> embedding(String textToEmbed) {
        return Uni.createFrom()
            .item(textToEmbed)
            .map(embeddingModel::embed)
            .map(Response::content)
            .map(Embedding::vectorAsList);
            
    }
```

Everything works, but the host I don't know if it makes sense. :thinking:

Reply:
I was expecting a `BlockingNotAllowedException` which was not thrown since the `Embedding` API is blocking. Mutiny is something I need to study more :smile:

Reply:
So from my understanding for now the only way to have this feature enabled is to force `Mutiny` to work on a worker thread, for all other scenarios some changes in `langchain4j` are needed.

```java
client.chatStreaming(request, version)
                .emitOn(Infrastructure.getDefaultWorkerPool())
                .subscribe()
```

Reply:
Yes, I guess for the time being, that's the only choice. It's quite unfortunate as it will reduce the concurrency and response time (well, the LLM is generally taking so much time that a thread switch is invisible).

Reply:
Maybe the virtual threads (the magic option) can ease this pain 😄

Reply:
> (well, the LLM is generally taking so much time that a thread switch is invisible)

This is actually true, the time taken to respond today is not low.

Reply:
> Maybe the virtual threads (the magic option) can ease this pain 😄

You need to make sure you read the small print :-)
But it's a good point. I'm wondering if we could change the dispatching strategy based on the annotation from the tool method.

For example:

```
@Tool 
String compute(...) {
}
```

This is an imperative method, we invoke it on the worker thread pool.

If the method is annotated with `@RunOnVirtualThread` like the following:

```
@Tool 
@RunOnVirtualThread
String compute(...) {
}
```

Instead of the regular worker thread pool, you would use the virtual thread scheduler (we expose it as a bean). 

It would be great in the future to support non-blocking tools returning `Uni<T>`, but not necessary something for today as it requires changing langchain4j.

Reply:
> You need to make sure you read the small print :-)

I know, I know, the magic doesn't always do what it's supposed to 😄 . 

> If the method is annotated with `@RunOnVirtualThread` like the following

Today it is not possible to annotate the tools method with `@RunOnVirtualThread`, but this is something that can be very useful. In this case the developer can create a worker thread or a virtual thread based on the business logic inside the method.